x/ibc-rate-limit/ibc_middleware_test.go

```diff
@@ -0,0 +1,351 @@
+	transfertypes "github.com/cosmos/ibc-go/v3/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v3/modules/core/02-client/types"
+	ibctesting "github.com/cosmos/ibc-go/v3/testing"
+	"github.com/osmosis-labs/osmosis/v12/app"
+	"github.com/osmosis-labs/osmosis/v12/app/apptesting"
+	"github.com/osmosis-labs/osmosis/v12/x/ibc-rate-limit/testutil"
+	"github.com/osmosis-labs/osmosis/v12/x/ibc-rate-limit/types"
+	"github.com/stretchr/testify/suite"
+)
+
+type MiddlewareTestSuite struct {
+	apptesting.KeeperTestHelper
+
+	coordinator *ibctesting.Coordinator
+
+	// testing chains used for convenience and readability
+	chainA *osmosisibctesting.TestChain
+	chainB *osmosisibctesting.TestChain
+	path   *ibctesting.Path
+}
+
+// Setup
+func TestMiddlewareTestSuite(t *testing.T) {
+	suite.Run(t, new(MiddlewareTestSuite))
+}
+
+func SetupTestingApp() (ibctesting.TestingApp, map[string]json.RawMessage) {
+	osmosisApp := app.Setup(false)
+	return osmosisApp, app.NewDefaultGenesisState()
+}
+
+func NewTransferPath(chainA, chainB *osmosisibctesting.TestChain) *ibctesting.Path {
+	path := ibctesting.NewPath(chainA.TestChain, chainB.TestChain)
+	path.EndpointA.ChannelConfig.PortID = ibctesting.TransferPort
+	path.EndpointB.ChannelConfig.PortID = ibctesting.TransferPort
+	path.EndpointA.ChannelConfig.Version = transfertypes.Version
+	path.EndpointB.ChannelConfig.Version = transfertypes.Version
+	return path
+}
+
+func (suite *MiddlewareTestSuite) SetupTest() {
+	suite.Setup()
+	ibctesting.DefaultTestingAppInit = SetupTestingApp
+	suite.coordinator = ibctesting.NewCoordinator(suite.T(), 2)
+	suite.chainA = &osmosisibctesting.TestChain{
+		TestChain: suite.coordinator.GetChain(ibctesting.GetChainID(1)),
+	}
+	// Remove epochs to prevent  minting
+	suite.chainA.MoveEpochsToTheFuture()
+	suite.chainB = &osmosisibctesting.TestChain{
+		TestChain: suite.coordinator.GetChain(ibctesting.GetChainID(2)),
+	}
+	suite.path = NewTransferPath(suite.chainA, suite.chainB)
+	suite.coordinator.Setup(suite.path)
+}
+
+// Helpers
+
+// NewValidMessage generates a new sdk.Msg of type MsgTransfer.
+// forward=true means that the message will be a "send" message, while forward=false is for  a "receive" message.
+// amount represents the amount transferred
+func (suite *MiddlewareTestSuite) NewValidMessage(forward bool, amount sdk.Int) sdk.Msg {
+	var coins sdk.Coin
+	var port, channel, accountFrom, accountTo string
+
+	if forward {
+		coins = sdk.NewCoin(sdk.DefaultBondDenom, amount)
+		port = suite.path.EndpointA.ChannelConfig.PortID
+		channel = suite.path.EndpointA.ChannelID
+		accountFrom = suite.chainA.SenderAccount.GetAddress().String()
+		accountTo = suite.chainB.SenderAccount.GetAddress().String()
+	} else {
+		coins = transfertypes.GetTransferCoin(
+			suite.path.EndpointB.ChannelConfig.PortID,
+			suite.path.EndpointB.ChannelID,
+			sdk.DefaultBondDenom,
+			sdk.NewInt(1),
+		)
+		coins = sdk.NewCoin(sdk.DefaultBondDenom, amount)
```

Comment:
why the overwrite and not:
```suggestion
		coins = transfertypes.GetTransferCoin(
			suite.path.EndpointB.ChannelConfig.PortID,
			suite.path.EndpointB.ChannelID,
			sdk.DefaultBondDenom,
			amount,
		)
```

Reply:
This got overlooked. Thanks for the catch!

Reply:
The correct code here is `coins = sdk.NewCoin(sdk.DefaultBondDenom, amount)` because that message is sent from chainB to chainA. What you said about getting the funds still applies though: I was not wrapping the token before getting the value and this was only working because both chains in my tests were the same chain.

I've fixed that in my latest commit (032f33b15feae8f40d3b088bd6d2bdb90d28a2bb)

Reply:
Yeah, this is testing native B tokens being sent to a. But maybe I should also add a test for native A tokens being sent "back" from B to A